python/google/protobuf/internal/enum_type_wrapper.py

```diff
@@ -45,6 +45,13 @@ class EnumTypeWrapper(object):
 
   DESCRIPTOR = None
 
+  # This is a type alias, which mypy typing stubs can type as
+  # a genericized parameter constrained to an int, allowing subclasses
+  # to be typed with more constraint
+  # Eg.
+  # def Name(self, number: MyGeneratedEnum.V) -> str
+  V = int
```

Comment:
I'd prefer a more explicit name like `ValueType`

Reply:
Hi - since this is for compatibility with https://github.com/dropbox/mypy-protobuf - this is a bit of a large non-backward compatible suggestion unfortunately.

It has the ergonomic downside of having very lengthy type names that don't end up being much more useful (IMO). In either case, you get an error from mypy of the form:
```
incompatible type "MyGeneratedEnum"; expected "ValueType"
incompatible type "MyGeneratedEnum"; expected "V"
```
both of which require some amount of looking inside the generated mypy-protobuf `.pyi` to understand. I believe there is a common confusion amongst protobuf users that `MyGeneratedEnum` is also an `int`, when protobuf generates it as an `EnumTypeWrapper`.

Here's an example `.proto` enum (OuterEnum) https://github.com/dropbox/mypy-protobuf/blob/master/proto/testproto/test.proto#L11
And the corresponding generated `.pyi` in mypy-protobuf `OuterEnum.V` https://github.com/dropbox/mypy-protobuf/blob/master/test/generated/testproto/test_pb2.pyi.expected#L25

Reply:
mypy-protobuf has this documented here https://github.com/dropbox/mypy-protobuf#types-enum-int-values-more-strongly.

If we do want to change the `V` to something else, it could be possible via a compatibility version supporting both `V` and `ValueType` types and eventually dropping support for `V`. I know Dropbox alone has thousands of instances of `V` already, as do other companies/codebases using mypy+protobuf. I'm hesitant to do such a change without some strong feedback + consensus - to avoid churning on name changes here.

Reply:
This is an unfortunate situation... and the approach looks a bit backwards to me: mypy should not require changes to Python or common libraries.

But if I got this correctly, all current usages need quotes, as in: `def f(x: 'MyEnum.V')` and this `V` is provided by mypy-protobuf generated code.
Since removing the quotes is a change anyway, the recommended end state could be `def f(x: MyEnum.ValueType)` (no quotes, but longer name), and the `V` would be supported by mypy as an alias (shorter, but with quotes)

Reply:
> This is an unfortunate situation... and the approach looks a bit backwards to me: mypy should not require changes to Python or common libraries.

Agreed.

The reason this comes up is because `mypy-protobuf` is attempting to get *stronger* typing than what protobuf itself generates at runtime. Protobuf generates all enum values as `int`. The stronger typing has proven to catch bugs and is one of the value adds of mypy-protobuf, but in order to namespace the `NewType("ValueType", int)` into `MyEnum`, it requires this change.

Starting with python 3.10 there is a new feature for postponed evaluation of annotations
https://www.python.org/dev/peps/pep-0563/
(or starting in python 3.7 with `from __future__ import annotations`)
However, there are still cases where this is insufficient to avoid evaluation - notably
```
from typing import cast
cast(MyEnum.ValueType, x)
```

> Since removing the quotes is a change anyway, the recommended end state could be def f(x: MyEnum.ValueType) (no quotes, but longer name), and the V would be supported by mypy as an alias (shorter, but with quotes)

I'm happy with your suggestion for a backward-compatible way to move toward a more verbose name. I can work with that in `mypy-protobuf`. @amauryfa - are you a review authority here? I can make the change. Just want to make sure I avoid discussion churn on these sorts of points.

Side note: There is an alternate python generated code option that sidestepped this problem entirely by generating enum values as enums https://github.com/danielgtaylor/python-betterproto. I've never tried it.

Reply:
Updated the diff with this idea of `s/V/ValueType`

Reply:
@perezd can you approve running to workflow again?